File: native-array-memory/prolog/interpreter.pl
%% author: 謝宇恆 / XIE Yuheng
%% github: https://github.com/xieyuheng

%% threaded code interpreter embeded in prolog

%% :- module(interpreter,
%%           [
%%            begin_to_interpret_threaded_code/0]).

:- use_module(array).

:- use_module(buffer).

%% any value can be stored in the cell_area
:- make_buffer(cell_area, 1024, 0).
data(Value) :-
  %: + value
  buffer_get_cursor(cell_area, Cursor),
  buffer_set(cell_area, Cursor, Value),
  buffer_add_cursor(cell_area, 1).

:- use_module(stack).

:- make_stack(argument_stack, 1024, 0).
:- make_stack(return_stack, 1024, 0).

%% no clean namespace of primitive_function

:- nb_setval(next_explainer_argument, 0).
next() :-
  stack_pop(return_stack, JOJO),
  buffer_get(cell_area, JOJO, JO),
  buffer_get(cell_area, JO, EXPLAINER),
  NEW_JOJO is JOJO + 1,
  stack_push(return_stack, NEW_JOJO),
  EXPLAINER_ARGUMENT is JO + 1,
  nb_setval(next_explainer_argument, EXPLAINER_ARGUMENT),
  call(EXPLAINER).

:- use_module(table).
:- make_table(name_jo_table, 1024).

primitive_function_explainer :-
  nb_getval(next_explainer_argument, EXPLAINER_ARGUMENT),
  buffer_get(cell_area, EXPLAINER_ARGUMENT, PRIMITIVE_FUNCTION),
  call(PRIMITIVE_FUNCTION).

declare_primitive_function(Name) :-
  buffer_get_cursor(cell_area, Cursor),
  table_set(name_jo_table, Name, Cursor),
  data(primitive_function_explainer),
  data(Name).


function_explainer :-
  nb_getval(next_explainer_argument, EXPLAINER_ARGUMENT),
  stack_push(return_stack, EXPLAINER_ARGUMENT),
  next.

define_function(Name, NameList) :-
  buffer_get_cursor(cell_area, Cursor),
  table_set(name_jo_table, Name, Cursor),
  data(function_explainer),
  define_body(NameList).

define_body([]).
define_body([Name|Rest]) :-
  table_find(name_jo_table, Name, JO, _),
  data(JO),
  define_body(Rest).


variable_explainer :-
  nb_getval(next_explainer_argument, EXPLAINER_ARGUMENT),
  buffer_get(cell_area, EXPLAINER_ARGUMENT, Value),
  stack_push(argument_stack, Value),
  next.

define_variable(Name, Value) :-
  buffer_get_cursor(cell_area, Cursor),
  table_set(name_jo_table, Name, Cursor),
  data(variable_explainer),
  data(Value).


push(Value) :-
  %: +value
  stack_push(argument_stack, Value).

pop(Value) :-
  %: value
  stack_pop(argument_stack, Value).


:- declare_primitive_function(end).
end :-
  stack_pop(return_stack, _),
  next.

:- declare_primitive_function(bye).
bye :-
  stack_pop(return_stack, _),
  writeln('bye bye ^-^/').

:- declare_primitive_function(dup).
dup :-
  pop(A),
  push(A),
  push(A),
  next.

:- declare_primitive_function(mul).
mul :-
  pop(B),
  pop(A),
  N is A * B,
  push(N),
  next.

:- declare_primitive_function(simple_wirte).
simple_wirte :-
  pop(A),
  writeln(A),
  next.

:- define_variable(little_test_number, 4).

:- define_function(
       square,
       [ dup,
         mul,
         end ]).

:- define_function(
       little_test,
       [ little_test_number,
         square,
         simple_wirte,
         bye ]).

:- define_function(
       first_function,
       [ little_test,
         end ]).


begin_to_interpret_threaded_code :-
  table_find(name_jo_table, first_function, JO, _),
  FUNCTION_BODY_FOR_LITTLE_TEST is JO + 1,
  stack_push(return_stack, FUNCTION_BODY_FOR_LITTLE_TEST),
  next.

:- begin_to_interpret_threaded_code.

:- halt.

File: native-array-memory/js-2/interpreter.js
"use strict";

let cell_area = new Buffer(1024 * 1024, 0);

let argument_stack = new Stack(1024, 0);
let return_stack = new Stack(1024, 0);

let primitive_function_record = new Buffer(1024, 0);

let create_primitive_function = (fun) => {
  let return_address = primitive_function_record.cursor;
  primitive_function_record
    .set(return_address, fun);
  primitive_function_record.add_cursor(1);
  return return_address;
};

var address_after_explainer = 0;

let interpreter = () => {
  try {
    while (true) {
      let jojo = return_stack.pop();
      let jo = cell_area.get(jojo);
      let explainer = cell_area.get(jo);
      return_stack.push(jojo + 1);
      address_after_explainer = jo + 1;
      primitive_function_record.get(explainer).call();
      continue;
    }

  } catch (string) {
    switch (string) {
    case "bye":
      break;
    }
  }
};

let in_host_tag_record = new Map();

let data = (value) => {
  cell_area.set(cell_area.cursor, value);
  cell_area.add_cursor(1);
};

let mark = (tag_string) => {
    in_host_tag_record.set(tag_string, cell_area.cursor);
};

let link = 0;

let define_header = (tag_string, explainer) => {
  data(link);
  link = cell_area.cursor - 1;
  mark(tag_string);
  data(explainer);
};

let primitive_function_explainer = create_primitive_function(
  () => {
    primitive_function_record.get(
      cell_area.get(address_after_explainer)
    ).call();
  });

let define_primitive_function = (tag_string, fun) => {
  let function_index = create_primitive_function(fun);
  define_header(tag_string, primitive_function_explainer);
  data(function_index);
};

let function_explainer = create_primitive_function(
  () => {
    return_stack.push(address_after_explainer);
  });

let define_function = (tag_string, function_tag_string_array) => {
  define_header(tag_string, function_explainer);
  function_tag_string_array.forEach(
    function_tag_string => {
      data(in_host_tag_record
           .get(function_tag_string));
    }
  );
};

let variable_explainer = create_primitive_function(
  () => {
    argument_stack.push(
      cell_area.get(address_after_explainer));
  });

let define_variable = (tag_string, value) => {
  define_header(tag_string, variable_explainer);
  data(value);
};

define_primitive_function(
  "end",
  () => {
    return_stack.pop();
  }
);

define_primitive_function(
  "bye",
  () => {
    console.log("bye bye ^-^/");
    throw "bye";
  }
);

define_primitive_function(
  "dup",
  () => {
    let a = argument_stack.pop();
    argument_stack.push(a);
    argument_stack.push(a);
  }
);

define_primitive_function(
  "mul",
  () => {
    let a = argument_stack.pop();
    let b = argument_stack.pop();
    argument_stack.push(a * b);
  }
);

define_primitive_function(
  "simple-wirte",
  () => {
    console.log(argument_stack.pop());
  }
);

define_variable("little-test-number", 4);

define_function(
  "square",
  [ "dup",
    "mul",
    "end"
  ]
);

define_function(
  "little-test",
  [ "little-test-number",
    "square",
    "simple-wirte",
    "bye"
  ]
);

define_function(
  "first-function",
  [ "little-test",
    "end"
  ]
);

let begin_to_interpret_threaded_code = () => {
  let function_body_for_little_test =
        in_host_tag_record.get("first-function") + 1;
  return_stack.push(function_body_for_little_test);
  interpreter();
};

begin_to_interpret_threaded_code();
